smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,235 @@
+
+            connectionInternal.asyncGo(() -> {
+
+                // Look for connection endpoints, the result obtained will contain the endpoints obtained in the lookup process.
+                Result<WebsocketRemoteConnectionEndpoint> result = RemoteWebsocketConnectionEndpoints.lookup(connectionInternal);
+
+                LookupConnectionEndpointsResult endpointsResult;
+
+                // If discoveredRemoteConnectionEndpoints list is empty set the result to WebsocketEndpointsDiscoveryFailed,
+                // or otherwise to DiscoveredWebsocketEndpoints.
+                if (result.discoveredRemoteConnectionEndpoints.isEmpty()) {
+                    endpointsResult = new WebsocketEndpointsDiscoveryFailed(result);
+                } else {
+                    endpointsResult = new DiscoveredWebsocketEndpoints(result);
+                }
+                websocketEndpointsLookupFuture.setResult(endpointsResult);
+            });
+
+            futures.add(websocketEndpointsLookupFuture);
+            return futures;
+        }
+
+        @Override
+        protected void loadConnectionEndpoints(LookupConnectionEndpointsSuccess lookupConnectionEndpointsSuccess) {
+            discoveredWebsocketEndpoints = (DiscoveredWebsocketEndpoints) lookupConnectionEndpointsSuccess;
+        }
+
+        @Override
+        protected void afterFiltersClosed() {
+        }
+
+        @Override
+        protected void disconnect() {
+            connectionAttemptState.websocket.disconnect();
+        }
+
+        @Override
+        protected void notifyAboutNewOutgoingElements() {
+            afterOutgoingElementsQueueModified();
+        }
+
+        @Override
+        public SSLSession getSslSession() {
+            return null;
+        }
+
+        @Override
+        public boolean isConnected() {
+            return connectionInternal.connection.isConnected();
+        }
+
+        @Override
+        public Stats getStats() {
+            return null;
+        }
+
+        public final class DiscoveredWebsocketEndpoints implements LookupConnectionEndpointsSuccess {
+            final RemoteWebsocketConnectionEndpoints.Result<WebsocketRemoteConnectionEndpoint> result;
+
+            DiscoveredWebsocketEndpoints(Result<WebsocketRemoteConnectionEndpoint> result) {
+                this.result = result;
+            }
+
+            public RemoteWebsocketConnectionEndpoints.Result<WebsocketRemoteConnectionEndpoint> getResult() {
+                return result;
+            }
+        }
+
+        final class WebsocketEndpointsDiscoveryFailed implements LookupConnectionEndpointsFailed {
+            final List<RemoteConnectionEndpointLookupFailure> lookupFailures;
+
+            WebsocketEndpointsDiscoveryFailed(
+                            RemoteWebsocketConnectionEndpoints.Result<WebsocketRemoteConnectionEndpoint> result) {
+                lookupFailures = result.lookupFailures;
```

Comment:
Nit: This allows the user to pass `null` as `result`.
Also, even though the list itself is declared `final` above, it is still possible to modify it, which is undesirable.